File: linkedlist/plusOneLinkedList.py
# https://leetcode.com/problems/plus-one-linked-list/


# Definition of ListNode:
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next


# 81 ms time cost.
# 6.00 MB memory cost.
# Your submission beats 84.60 % Submissions.
class Solution:
    def plusOne(self, head: ListNode) -> ListNode:
        def __reverse(head: ListNode) -> ListNode:
            if head is None or head.next is None:
                return head
            new_head = __reverse(head.next)
            head.next.next = head
            head.next = None
            return new_head

        sum, carry = 0, 1
        new_head = __reverse(head)
        current_node, prev_node = new_head, new_head
        while current_node is not None:
            sum = carry + current_node.val
            current_node.val = sum % 10
            carry = sum // 10
            prev_node = current_node
            current_node = current_node.next
        if carry > 0:
            prev_node.next = ListNode(carry)
        head = __reverse(new_head)

        return head

File: linkedlist/PlusOneLinkedList.java
package linkedlist;

// https://leetcode.com/problems/plus-one-linked-list/

public class PlusOneLinkedList {
    public ListNode addToNum(ListNode head) {
        int sum = 0, carry = 1;
        ListNode newHead = reverse(head);
        ListNode currentNode = newHead, prevNode = newHead;
        while (currentNode != null) {
            sum = carry + currentNode.val;
            currentNode.val = sum % 10;
            carry = sum / 10;
            prevNode = currentNode;
            currentNode = currentNode.next;
        }
        if (carry > 0)
            prevNode.next = new ListNode(carry);
        head = reverse(newHead);
        return head;
    }

    private ListNode reverse(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode newHead = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }
}
